networking/decentralized-firewall/validator/validator.py

```diff
@@ -0,0 +1,262 @@
+        return is_ok
+
+
+class NetworkPorts(Validator):
+    """ Custom ports validator """
+    tag = 'networkports'
+    settings = {}
+    mode = None
+    _type = None
+    allowed_port_map = []
+    approved_port_map = []
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        for port in self.settings['allowedPorts']:
+            ports = self._process_port_definition(port['ports'])
+            self.allowed_port_map.extend(ports)
+            if port['approved']:
+                self.approved_port_map.extend(ports)
+
+    def _process_port_definition(self, port_definition):
+        ports = []
+        if not isinstance(port_definition, int) and '-' in port_definition:
+            start, end = port_definition.split('-', 2)
+            for port in range(int(start), int(end) + 1):
+                ports.append(int(port))
+        else:
+            ports.append(int(port_definition))
+        return ports
+
+    def fail(self, value):
+        mode_str = 'allowed'
+        if self.mode == 'approve':
+            mode_str = 'automatically approved'
+        return '\'%s\' is not an %s IP port.' % (value, mode_str)
+
+    def _is_valid(self, value):
+        ports = self._process_port_definition(value)
+        is_ok = True
+        for port in ports:
+            if self.mode == 'approve' and port not in self.approved_port_map:
+                is_ok = False
+                break
+            elif port not in self.allowed_port_map:
+                is_ok = False
+                break
+
+        return is_ok
+
+
+class FirewallValidator:
+    schema = None
+    settings = None
+    validators = None
+
+    def __init__(self, settings, mode):
+        self.settings = settings
+
+        self.validators = DefaultValidators.copy()
+        Netmask.settings = self.settings
+        Netmask.mode = mode
+        self.validators[Netmask.tag] = Netmask
+
+        NetworkTag.settings = self.settings
+        NetworkTag.mode = mode
+        self.validators[NetworkTag.tag] = NetworkTag
+
+        ServiceAccount.settings = self.settings
+        ServiceAccount.mode = mode
+        self.validators[ServiceAccount.tag] = ServiceAccount
+
+        NetworkPorts.settings = self.settings
+        NetworkPorts.mode = mode
+        self.validators[NetworkPorts.tag] = NetworkPorts
+
+    def set_schema_from_file(self, schema):
+        self.schema = yamale.make_schema(path=schema,
+                                         validators=self.validators)
+
+    def set_schema_from_string(self, schema):
+        self.schema = yamale.make_schema(content=schema,
+                                         validators=self.validators)
+
+    def validate_file(self, file):
+        print('Validating %s...' % (file), file=sys.stderr)
+        data = yamale.make_data(file)
+        yamale.validate(self.schema, data)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Validate firewall rules.')
```

Comment:
Can you break out the bulk of this code in a `main` function? We use `click` in other tools in this repo, would you mind switching to it instead of using argparse?

Reply:
Done.